File: backend-ruby/app/controllers/api/v1/phong_thi_controller.rb
class Api::V1::PhongThiController < ApplicationController
  before_action :set_phong_thi, only: [:show, :update, :destroy]

  def index
    records = PhongThi.order(PhongThiID: :asc)
    render json: records, status: :ok
  end

  def show
    render json: @phong_thi, status: :ok
  end

  def create
    record = PhongThi.new(phong_thi_params)
    if record.save
      render json: record, status: :created
    else
      render json: { errors: record.errors.full_messages }, status: :unprocessable_entity
    end
  end

  def update
    if @phong_thi.update(phong_thi_params)
      render json: @phong_thi, status: :ok
    else
      render json: { errors: @phong_thi.errors.full_messages }, status: :unprocessable_entity
    end
  end

  def destroy
    @phong_thi.destroy
    render json: { message: "Deleted exam room with ID #{@phong_thi.PhongThiID}." }, status: :ok
  rescue ActiveRecord::InvalidForeignKey
    render json: { error: "Cannot delete exam room with ID #{@phong_thi.PhongThiID} because it is used by an exam." }, status: :conflict
  end

  private

  def set_phong_thi
    @phong_thi = PhongThi.find_by(PhongThiID: params[:id])
    unless @phong_thi
      render json: { error: "Exam room with ID #{params[:id]} was not found." }, status: :not_found
    end
  end

  def phong_thi_params
    params.require(:phong_thi).permit(:MaPhong, :TenPhong, :ToaNha, :Tang, :SucChua, :TrangThai, :SoHang, :SoCot)
  end
end

File: backend-csharp/Controllers/PhongThiController.cs
using System.Text.Json;
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/phong_thi")]
public class PhongThiController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpGet]
    public async Task<IActionResult> Index() => Ok(await dbContext.PhongThi.OrderBy(record => record.PhongThiID).ToListAsync());

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        PhongThi? record = await dbContext.PhongThi.FindAsync(id);
        return record is null ? NotFound(Error($"Exam room with ID {id} was not found.")) : Ok(record);
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] JsonElement body)
    {
        PhongThi? record = ReadBody<PhongThi>(body, "phong_thi");
        if (record is null)
        {
            return UnprocessableEntity(Errors("Invalid exam room payload."));
        }

        dbContext.PhongThi.Add(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return CreatedAtAction(nameof(Show), new { id = record.PhongThiID }, record);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Duplicate room code or invalid room data."));
        }
    }

    [HttpPatch("{id:int}")]
    [HttpPut("{id:int}")]
    public async Task<IActionResult> Update(int id, [FromBody] JsonElement body)
    {
        PhongThi? record = await dbContext.PhongThi.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error($"Exam room with ID {id} was not found."));
        }

        PhongThi? payload = ReadBody<PhongThi>(body, "phong_thi");
        if (payload is null)
        {
            return UnprocessableEntity(Errors("Invalid exam room payload."));
        }

        record.MaPhong = payload.MaPhong;
        record.TenPhong = payload.TenPhong;
        record.ToaNha = payload.ToaNha;
        record.Tang = payload.Tang;
        record.SucChua = payload.SucChua;
        record.SoHang = payload.SoHang;
        record.SoCot = payload.SoCot;
        record.TrangThai = payload.TrangThai;
        record.CapNhatLuc = DateTime.Now;
        await dbContext.SaveChangesAsync();
        return Ok(record);
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Destroy(int id)
    {
        PhongThi? record = await dbContext.PhongThi.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error($"Exam room with ID {id} was not found."));
        }

        dbContext.PhongThi.Remove(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return Ok(new { message = $"Deleted exam room with ID {record.PhongThiID}." });
        }
        catch (DbUpdateException)
        {
            return Conflict(Error($"Cannot delete exam room with ID {record.PhongThiID} because it is used by an exam."));
        }
    }
}
